stream/stream.js

```diff
@@ -2,161 +2,136 @@
+	stream.changing = function() {
+		open(stream) && (stream.state = "changing")
+		dependentStreams.forEach(function(s) {
+			s.dependent && s.dependent.changing()
+			s.changing()
+		})
 	}
-	initStream(stream)
 
-	if (arguments.length > 0 && arguments[0] !== HALT) updateStream(stream, arguments[0])
+	stream.state = arguments.length && value !== Stream.SKIP ? "active" : "pending"
 
-	return stream
-}
-function initStream(stream) {
-	stream.constructor = createStream
-	stream._state = {id: guid++, value: undefined, state: 0, derive: undefined, recover: undefined, deps: {}, parents: [], endStream: undefined, unregister: undefined}
-	stream.map = stream["fantasy-land/map"] = map, stream["fantasy-land/ap"] = ap, stream["fantasy-land/of"] = createStream
-	stream.toJSON = toJSON
-
-	Object.defineProperties(stream, {
-		end: {get: function() {
-			if (!stream._state.endStream) {
-				var endStream = createStream()
-				endStream.map(function(value) {
-					if (value === true) {
-						unregisterStream(stream)
-						endStream._state.unregister = function(){unregisterStream(endStream)}
-					}
-					return value
-				})
-				stream._state.endStream = endStream
+	stream.map = function(fn, ignoreInitial) {
+		var target = stream.state === "active" && ignoreInitial !== Stream.SKIP
+			? Stream(fn(value))
+			: Stream()
+
+		dependentStreams.push(target)
+		dependentFns.push(fn)
+		return target
+	}
+
+	let end
+	function createEnd() {
+		end = Stream()
+		end.map(function(value) {
+			if (value === true) {
+				stream.state = "ended"
+				dependentStreams.length = dependentFns.length = 0
 			}
-			return stream._state.endStream
-		}}
-	})
-}
-function updateStream(stream, value) {
-	updateState(stream, value)
-	for (var id in stream._state.deps) updateDependency(stream._state.deps[id], false)
-	if (stream._state.unregister != null) stream._state.unregister()
-	finalize(stream)
-}
-function updateState(stream, value) {
-	stream._state.value = value
-	stream._state.changed = true
-	if (stream._state.state !== 2) stream._state.state = 1
-}
-function updateDependency(stream, mustSync) {
-	var state = stream._state, parents = state.parents
-	if (parents.length > 0 && parents.every(active) && (mustSync || parents.some(changed))) {
-		var value = stream._state.derive()
-		if (value === HALT) return unregisterStream(stream)
-		updateState(stream, value)
+			return value
+		})
+		return end
 	}
-}
-function finalize(stream) {
-	stream._state.changed = false
-	for (var id in stream._state.deps) stream._state.deps[id]._state.changed = false
-}
 
-function combine(fn, streams) {
-	if (!streams.every(valid)) throw new Error("Ensure that each item passed to stream.combine/stream.merge is a stream")
-	return initDependency(createStream(), streams, function() {
-		return fn.apply(this, streams.concat([streams.filter(changed)]))
-	})
-}
+	stream.toJSON = function() { return value != null && typeof value.toJSON === "function" ? value.toJSON() : value }
```

Comment:
Nice toJSON method, I just might copy that to some of my projects!

I'm however wondering whether it might be a good idea to add an `inspect` method as well, to help people while using the devTools.

It could be simple to implement, e.g.

```js
stream.inspect = function() { return 'Stream(' + stream.toJSON() +  ')'; }
```

Or something along those lines.

That would allow people to just however over a stream while debugging to see that value contained inside.

Reply:
Yeah, toString was removed due to issues with the diff - https://github.com/MithrilJS/mithril.js/issues/2150 , but inspect would be great.. Do you have some info on what browsers support that? (can't make it work in chrome)

Reply:
hmm... it used to work in chrome, maybe it's been removed :confused:

Reply:
oh.. yeah I can't seem to find much by googling... Did stumble over [this](http://www.mattzeunert.com/2016/02/19/custom-chrome-devtools-object-formatters.html) though, but that seems a bit much...

Reply:
Seems a bit much, `inspect` seems to be a Node only thing: https://nodejs.org/dist/latest-v4.x/docs/api/util.html#util_custom_inspect_function_on_objects